// @segment/consent-manager/esm/consent-manager-builder/preferences.js
// TODO: remove duplicate cookie library from bundle
import cookies from 'js-cookie';
import topDomain from '@segment/top-domain';

var COOKIE_KEY = 'tracking-preferences';
var COOKIE_EXPIRES = 365;

// TODO: harden against invalid cookies
export function loadPreferences() {
  var preferences = cookies.getJSON(COOKIE_KEY);

  if (!preferences) {
    return {};
  }

  return {
    destinationPreferences: preferences.destinations,
    customPreferences: preferences.custom
  };
}

export function savePreferences(_ref) {
  var destinationPreferences = _ref.destinationPreferences,
      customPreferences = _ref.customPreferences,
      cookieDomain = _ref.cookieDomain;

  window.analytics.identify({
    destinationTrackingPreferences: destinationPreferences,
    customTrackingPreferences: customPreferences
  });

  var domain = cookieDomain || topDomain(window.location.href);
  var value = {
    version: 1,
    destinations: destinationPreferences,
    custom: customPreferences
  };
  cookies.set(COOKIE_KEY, value, {
    expires: COOKIE_EXPIRES,
    domain: domain
  });
}
