// @segment/consent-manager/esm/consent-manager/preference-dialog.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _taggedTemplateLiteral from 'babel-runtime/helpers/taggedTemplateLiteral';

var _templateObject = _taggedTemplateLiteral(['\n  @media (max-width: 600px) {\n    display: none;\n  }\n'], ['\n  @media (max-width: 600px) {\n    display: none;\n  }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n  overflow-x: auto;\n  margin-top: 16px;\n'], ['\n  overflow-x: auto;\n  margin-top: 16px;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n  border-collapse: collapse;\n  font-size: 12px;\n'], ['\n  border-collapse: collapse;\n  font-size: 12px;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n  background: #f7f8fa;\n  color: #1f4160;\n  font-weight: 600;\n  text-align: left;\n  border-width: 2px;\n'], ['\n  background: #f7f8fa;\n  color: #1f4160;\n  font-weight: 600;\n  text-align: left;\n  border-width: 2px;\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n  font-weight: normal;\n  text-align: left;\n'], ['\n  font-weight: normal;\n  text-align: left;\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n  th,\n  td {\n    vertical-align: top;\n    padding: 8px 12px;\n    border: 1px solid rgba(67, 90, 111, 0.114);\n  }\n  td {\n    border-top: none;\n  }\n'], ['\n  th,\n  td {\n    vertical-align: top;\n    padding: 8px 12px;\n    border: 1px solid rgba(67, 90, 111, 0.114);\n  }\n  td {\n    border-top: none;\n  }\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n  input {\n    vertical-align: middle;\n  }\n  label {\n    display: block;\n    margin-bottom: 4px;\n    white-space: nowrap;\n  }\n'], ['\n  input {\n    vertical-align: middle;\n  }\n  label {\n    display: block;\n    margin-bottom: 4px;\n    white-space: nowrap;\n  }\n']);

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import Dialog from './dialog';
import { DefaultButton, GreenButton } from './buttons';

var hideOnMobile = /*#__PURE__*/css(_templateObject, 'label:hideOnMobile;');

var TableScroll = /*#__PURE__*/styled('div', {
  target: 'e1v3fxkj0',
  label: 'TableScroll'
})(_templateObject2);

var Table = /*#__PURE__*/styled('table', {
  target: 'e1v3fxkj1',
  label: 'Table'
})(_templateObject3);

var ColumnHeading = /*#__PURE__*/styled('th', {
  target: 'e1v3fxkj2',
  label: 'ColumnHeading'
})(_templateObject4);

var RowHeading = /*#__PURE__*/styled('th', {
  target: 'e1v3fxkj3',
  label: 'RowHeading'
})(_templateObject5);

var Row = /*#__PURE__*/styled('tr', {
  target: 'e1v3fxkj4',
  label: 'Row'
})(_templateObject6);

var InputCell = /*#__PURE__*/styled('td', {
  target: 'e1v3fxkj5',
  label: 'InputCell'
})(_templateObject7);

var PreferenceDialog = function (_PureComponent) {
  _inherits(PreferenceDialog, _PureComponent);

  function PreferenceDialog() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, PreferenceDialog);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = PreferenceDialog.__proto__ || _Object$getPrototypeOf(PreferenceDialog)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (e) {
      var onChange = _this.props.onChange;


      onChange(e.target.name, e.target.value === 'true');
    }, _this.handleSubmit = function (e) {
      var _this$props = _this.props,
          onSave = _this$props.onSave,
          marketingAndAnalytics = _this$props.marketingAndAnalytics,
          advertising = _this$props.advertising,
          functional = _this$props.functional;


      e.preventDefault();

      // Safe guard against browsers that don't prevent the
      // submission of invalid forms (Safari < 10.1)
      if (marketingAndAnalytics === null || advertising === null || functional === null) {
        return;
      }

      onSave();
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(PreferenceDialog, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          innerRef = _props.innerRef,
          onCancel = _props.onCancel,
          marketingDestinations = _props.marketingDestinations,
          advertisingDestinations = _props.advertisingDestinations,
          functionalDestinations = _props.functionalDestinations,
          marketingAndAnalytics = _props.marketingAndAnalytics,
          advertising = _props.advertising,
          functional = _props.functional,
          title = _props.title,
          content = _props.content;


      var buttons = React.createElement(
        'div',
        null,
        React.createElement(
          DefaultButton,
          { type: 'button', onClick: onCancel },
          'Cancel'
        ),
        React.createElement(
          GreenButton,
          { type: 'submit' },
          'Save'
        )
      );

      return React.createElement(
        Dialog,
        {
          innerRef: innerRef,
          title: title,
          buttons: buttons,
          onCancel: onCancel,
          onSubmit: this.handleSubmit
        },
        content,
        React.createElement(
          TableScroll,
          null,
          React.createElement(
            Table,
            null,
            React.createElement(
              'thead',
              null,
              React.createElement(
                Row,
                null,
                React.createElement(
                  ColumnHeading,
                  { scope: 'col' },
                  'Allow'
                ),
                React.createElement(
                  ColumnHeading,
                  { scope: 'col' },
                  'Category'
                ),
                React.createElement(
                  ColumnHeading,
                  { scope: 'col' },
                  'Purpose'
                ),
                React.createElement(
                  ColumnHeading,
                  { scope: 'col', className: hideOnMobile },
                  'Tools'
                )
              )
            ),
            React.createElement(
              'tbody',
              null,
              React.createElement(
                Row,
                null,
                React.createElement(
                  InputCell,
                  null,
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'functional',
                      value: 'true',
                      checked: functional === true,
                      onChange: this.handleChange,
                      'aria-label': 'Allow functional tracking',
                      required: true
                    }),
                    ' ',
                    'Yes'
                  ),
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'functional',
                      value: 'false',
                      checked: functional === false,
                      onChange: this.handleChange,
                      'aria-label': 'Disallow functional tracking',
                      required: true
                    }),
                    ' ',
                    'No'
                  )
                ),
                React.createElement(
                  RowHeading,
                  { scope: 'row' },
                  'Functional'
                ),
                React.createElement(
                  'td',
                  null,
                  React.createElement(
                    'p',
                    null,
                    'To monitor the performance of our site and to enhance your browsing experience.'
                  ),
                  React.createElement(
                    'p',
                    { className: hideOnMobile },
                    'For example, these tools enable you to communicate with us via live chat.'
                  )
                ),
                React.createElement(
                  'td',
                  { className: hideOnMobile },
                  functionalDestinations.map(function (d) {
                    return d.name;
                  }).join(', ')
                )
              ),
              React.createElement(
                Row,
                null,
                React.createElement(
                  InputCell,
                  null,
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'marketingAndAnalytics',
                      value: 'true',
                      checked: marketingAndAnalytics === true,
                      onChange: this.handleChange,
                      'aria-label': 'Allow marketing and analytics tracking',
                      required: true
                    }),
                    ' ',
                    'Yes'
                  ),
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'marketingAndAnalytics',
                      value: 'false',
                      checked: marketingAndAnalytics === false,
                      onChange: this.handleChange,
                      'aria-label': 'Disallow marketing and analytics tracking',
                      required: true
                    }),
                    ' ',
                    'No'
                  )
                ),
                React.createElement(
                  RowHeading,
                  { scope: 'row' },
                  'Marketing and Analytics'
                ),
                React.createElement(
                  'td',
                  null,
                  React.createElement(
                    'p',
                    null,
                    'To understand user behavior in order to provide you with a more relevant browsing experience or personalize the content on our site.'
                  ),
                  React.createElement(
                    'p',
                    { className: hideOnMobile },
                    'For example, we collect information about which pages you visit to help us present more relevant information.'
                  )
                ),
                React.createElement(
                  'td',
                  { className: hideOnMobile },
                  marketingDestinations.map(function (d) {
                    return d.name;
                  }).join(', ')
                )
              ),
              React.createElement(
                Row,
                null,
                React.createElement(
                  InputCell,
                  null,
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'advertising',
                      value: 'true',
                      checked: advertising === true,
                      onChange: this.handleChange,
                      'aria-label': 'Allow advertising tracking',
                      required: true
                    }),
                    ' ',
                    'Yes'
                  ),
                  React.createElement(
                    'label',
                    null,
                    React.createElement('input', {
                      type: 'radio',
                      name: 'advertising',
                      value: 'false',
                      checked: advertising === false,
                      onChange: this.handleChange,
                      'aria-label': 'Disallow advertising tracking',
                      required: true
                    }),
                    ' ',
                    'No'
                  )
                ),
                React.createElement(
                  RowHeading,
                  { scope: 'row' },
                  'Advertising'
                ),
                React.createElement(
                  'td',
                  null,
                  React.createElement(
                    'p',
                    null,
                    'To personalize and measure the effectiveness of advertising on our site and other websites.'
                  ),
                  React.createElement(
                    'p',
                    { className: hideOnMobile },
                    'For example, we may serve you a personalized ad based on the pages you visit on our site.'
                  )
                ),
                React.createElement(
                  'td',
                  { className: hideOnMobile },
                  advertisingDestinations.map(function (d) {
                    return d.name;
                  }).join(', ')
                )
              ),
              React.createElement(
                Row,
                null,
                React.createElement(
                  'td',
                  null,
                  'N/A'
                ),
                React.createElement(
                  RowHeading,
                  { scope: 'row' },
                  'Essential'
                ),
                React.createElement(
                  'td',
                  null,
                  React.createElement(
                    'p',
                    null,
                    'We use browser cookies that are necessary for the site to work as intended.'
                  ),
                  React.createElement(
                    'p',
                    null,
                    'For example, we store your website data collection preferences so we can honor them if you return to our site. You can disable these cookies in your browser settings but if you do the site may not work as intended.'
                  )
                ),
                React.createElement('td', { className: hideOnMobile })
              )
            )
          )
        )
      );
    }
  }]);

  return PreferenceDialog;
}(PureComponent);

PreferenceDialog.displayName = 'PreferenceDialog';
PreferenceDialog.propTypes = {
  innerRef: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  marketingDestinations: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired
  })).isRequired,
  advertisingDestinations: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired
  })).isRequired,
  functionalDestinations: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired
  })).isRequired,
  marketingAndAnalytics: PropTypes.bool,
  advertising: PropTypes.bool,
  functional: PropTypes.bool,
  title: PropTypes.node.isRequired,
  content: PropTypes.node.isRequired
};
PreferenceDialog.defaultProps = {
  marketingAndAnalytics: null,
  advertising: null,
  functional: null
};
export default PreferenceDialog;
