// ts-migrate-plugins/build/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSourceText = exports.tsIgnorePlugin = exports.stripTSIgnorePlugin = exports.reactShapePlugin = exports.reactPropsPlugin = exports.reactDefaultPropsPlugin = exports.reactClassStatePlugin = exports.reactClassLifecycleMethodsPlugin = exports.memberAccessibilityPlugin = exports.jsDocPlugin = exports.hoistClassStaticsPlugin = exports.explicitAnyPlugin = exports.eslintFixPlugin = exports.declareMissingClassPropertiesPlugin = exports.addConversionsPlugin = void 0;
const add_conversions_1 = __importDefault(require("./plugins/add-conversions"));
exports.addConversionsPlugin = add_conversions_1.default;
const declare_missing_class_properties_1 = __importDefault(require("./plugins/declare-missing-class-properties"));
exports.declareMissingClassPropertiesPlugin = declare_missing_class_properties_1.default;
const eslint_fix_1 = __importDefault(require("./plugins/eslint-fix"));
exports.eslintFixPlugin = eslint_fix_1.default;
const explicit_any_1 = __importDefault(require("./plugins/explicit-any"));
exports.explicitAnyPlugin = explicit_any_1.default;
const hoist_class_statics_1 = __importDefault(require("./plugins/hoist-class-statics"));
exports.hoistClassStaticsPlugin = hoist_class_statics_1.default;
const jsdoc_1 = __importDefault(require("./plugins/jsdoc"));
exports.jsDocPlugin = jsdoc_1.default;
const member_accessibility_1 = __importDefault(require("./plugins/member-accessibility"));
exports.memberAccessibilityPlugin = member_accessibility_1.default;
const react_class_lifecycle_methods_1 = __importDefault(require("./plugins/react-class-lifecycle-methods"));
exports.reactClassLifecycleMethodsPlugin = react_class_lifecycle_methods_1.default;
const react_class_state_1 = __importDefault(require("./plugins/react-class-state"));
exports.reactClassStatePlugin = react_class_state_1.default;
const react_default_props_1 = __importDefault(require("./plugins/react-default-props"));
exports.reactDefaultPropsPlugin = react_default_props_1.default;
const react_props_1 = __importDefault(require("./plugins/react-props"));
exports.reactPropsPlugin = react_props_1.default;
const react_shape_1 = __importDefault(require("./plugins/react-shape"));
exports.reactShapePlugin = react_shape_1.default;
const strip_ts_ignore_1 = __importDefault(require("./plugins/strip-ts-ignore"));
exports.stripTSIgnorePlugin = strip_ts_ignore_1.default;
const ts_ignore_1 = __importDefault(require("./plugins/ts-ignore"));
exports.tsIgnorePlugin = ts_ignore_1.default;
const updateSourceText_1 = __importDefault(require("./utils/updateSourceText"));
exports.updateSourceText = updateSourceText_1.default;
